Add support for ed25519 (2/2)

gjgd opened the pull request:
Follow up to https://github.com/decentralized-identity/ion-sdk/pull/20

Adding support for Ed25519 in IonSDK

gjgd committed:
Add validateEd25519OperationKey

gjgd committed:
Add test for validateEd25519OperationKey

gjgd committed:
Add support for Ed25519 in LocalSigner

gjgd committed:
Add tests for IonRequest with Ed25519 keys

gjgd committed:
Add validateOperationKey method to InputValidator

gjgd committed:
Add support for Ed25519 in IonDid

gjgd committed:
Add test for LocalSigner

gjgd committed:
Add more tests for InputValidator

OR13 commented on lib/InputValidator.ts:
```diff
@@ -1,14 +1,29 @@
  */
 export default class InputValidator {
+  /**
+   * Validates the schema of a ES256K JWK key.
```
```suggestion
   * Validates the schema of a Ed25519 or secp256k1 JWK
```

OR13 commented on lib/InputValidator.ts:
```diff
@@ -1,14 +1,29 @@
+    } else if (IonKey.isJwkEd25519(operationKeyJwk)) {
+      InputValidator.validateEd25519OperationKey(operationKeyJwk, operationKeyType);
+    } else {
+      throw new IonError(ErrorCode.UnsupportedKeyType, `JWK key should be Es256k or Ed25119.`);
```
```suggestion
      throw new IonError(ErrorCode.UnsupportedKeyType, `JWK key should be secp256k1 or Ed25519.`);
```

OR13 replied:
alternatively, refering to the JWS alg instead of the crv.... in that case `ES256K` and `EdDSA`...I find the confusion regarding this in sidetree frustrating...

OR13 commented on lib/IonDid.ts:
```diff
@@ -15,8 +16,8 @@ export default class IonDid {
   public static createLongFormDid (input: {
-    recoveryKey: JwkEs256k;
-    updateKey: JwkEs256k;
+    recoveryKey: JwkEs256k | JwkEd25519;
```
consider moving this or type to a proper type for `SidetreeOperationalPrivateKeyJwk` or similar

gjgd replied:
Added SidetreeKeyJwk type https://github.com/decentralized-identity/ion-sdk/pull/23/commits/bd6625238dc73c9f6ac5962cd3b9d22a87da588a

OR13 commented on lib/LocalSigner.ts:
```diff
@@ -12,30 +17,47 @@ export default class LocalSigner implements ISigner {
+        privateKeyJwk: this.privateKey
+      } as any);
+      const signer = key.signer();
+      const jwsSigner = await JWS.createSigner(signer, 'EdDSA', {
```
it might be worth a sanity test with https://github.com/panva/jose in this library to cover JWS interop, we have one here, but you can never test this kind of thing enough...  https://github.com/transmute-industries/verifiable-data/blob/main/packages/jose-ld/src/JWS/jose.test.ts

OR13 commented on lib/LocalSigner.ts:
```diff
@@ -12,30 +17,47 @@ export default class LocalSigner implements ISigner {
+      const compactJws = await jwsSigner.sign({ data: content });
+      return compactJws;
+    } else {
+      throw new IonError(ErrorCode.UnsupportedKeyType, `JWK key should be Es256k or Ed25119.`);
```
again, `alg` vs `crv` issue.

gjgd replied:
Updated https://github.com/decentralized-identity/ion-sdk/pull/23/commits/9825c9bbd6c156a777d5082f16b52e13cba87dc7

OR13 commented on tests/IonDid.spec.ts:
```diff
@@ -32,6 +35,23 @@ describe('IonDid', async () => {
+
+      const longFormDid = IonDid.createLongFormDid({ recoveryKey, updateKey, document });
+
+      const expectedMethodSpecificId = 'did:ion:EiBks49Ah-ZxE1-6Se-PUDk_4_ffylpeqmfMCWUBwwTM-g:eyJkZWx0YSI6eyJwYXRjaGVzIjpbeyJhY3Rpb24iOiJyZXBsYWNlIiwiZG9jdW1lbnQiOnsicHVibGljS2V5cyI6W3siaWQiOiJwdWJsaWNLZXlNb2RlbEVkMjU1MTkiLCJwdWJsaWNLZXlKd2siOnsiY3J2IjoiRWQyNTUxOSIsImt0eSI6Ik9LUCIsIngiOiJoQ3hhTVI2TlVPWEkxQzd3Nlh6MW1jaGMyd1M4RlZ2WGhuZ3NnWjBodHNZIn0sInB1cnBvc2VzIjpbImF1dGhlbnRpY2F0aW9uIiwia2V5QWdy ...
```
the first Long Form ION DID to support EdDSA / ED25519 ?!?!?!?!?! :)

OR13 reviewed (changes requested):
Main issue is with `alg` vs `crv` in the variable names... @thehenrytsai how should this be solved? normalize to `alg` ? in other others `EdDSA` everywhere you see `Ed25519` ?

gjgd committed:
Update lib/InputValidator.ts

Co-authored-by: Orie Steele <orie@or13.io>

gjgd committed:
Update lib/InputValidator.ts

Co-authored-by: Orie Steele <orie@or13.io>

gjgd committed:
Add SidetreeKeyJwk key type

gjgd committed:
Updated error message

gjgd commented:
@OR13 I addressed your comments. I am not so sure how to deal with the crv vs alg problem because Ed25519 is the name of the digital signature algorithm, so I think it does belong next to ES256k.

OR13 commented:
@gjgd https://www.iana.org/assignments/jose/jose.xhtml

@thehenrytsai @csuwildcat thoughts: 

```
EdDSA | EdDSA signature algorithms | alg
-- | -- | --
```

```
Ed25519 | Ed25519 signature algorithm key pairs | Optional
-- | -- | --
```

```
ES256K | ECDSA using secp256k1 curve and SHA-256
-- | --
```

```
secp256k1 | SECG secp256k1 curve | Optional
-- | -- | --
```

gjgd commented:
![Screenshot 2022-01-24 at 13 55 34](https://user-images.githubusercontent.com/7913565/150855065-b6d5bab8-74e4-4971-b5a0-848404c254c1.png)

This is unfortunate naming. Pretty sure Ed25519 is the name of the digital signature algorithm and it should have been `Curve25519` as the name of the associated curve.

ES256k is the implementation of the ECDSA signature algorithm that uses the secp256k1 curve
Ed25519 is the implementation of the EdDSA signature algorithm that uses the Curve25519 curve

Edit: from https://en.wikipedia.org/wiki/EdDSA
> Ed25519 is the EdDSA signature scheme using SHA-512 (SHA-2) and Curve25519

OR13 commented:
@gjgd you are correct, but ION only uses JWS algs... not raw signatures... hence the correct term in ION for a signature is a JWS alg... if ION supported raw ECDSA or EdDSA without a JOSE header this would be a different story.

thehenrytsai commented:
Thanks @gjgd and @OR13. I am now aware of the PR, will try take a look tomorrow.

thehenrytsai commented:
Forgive me if the question is silly, but shouldn't the reference implementation be updated first to support ED25519 before SDK claims to have the support? Else anyone using the SDK might be stuck with their DID document?

gjgd commented:
Planning to use the ion sdk in a project that's not the reference implementation. I assume the reference implementation would throw an error for keys that are not supported?

csuwildcat commented:
> Planning to use the ion sdk in a project that's not the reference implementation. I assume the reference implementation would throw an error for keys that are not supported?

If you mean the actual ION node code, yes, that will throw if it has not been updated and a release is not propagated across the community in a network upgrade.

gjgd commented:
I can add a `console.warning` indicating that an ion node may reject an operation with ed25519 keys until the reference implementation supports it, would that work?

csuwildcat commented:
> I can add a `console.warning` indicating that an ion node may reject an operation with ed25519 keys until the reference implementation supports it, would that work?

There's certainly nothing wrong with Sidetree in general supporting Ed25519, and I'd even like to see ION support it, but my concern is that if we allow people to create long-form IDs in a way that isn't actually valid at the network/protocol level, it will cause strange issues. The best thing would be to see the node code updated with matching support so it actually does work, then push them out together. What's the plan on the node implementation side?

gjgd commented:
I understand the concerns and the potential confusion this could create, but are you suggesting that we cannot merge this PR (that is already approved) until Sidetree nodes support Ed25519?

If not, please suggest a set of changes that would alleviate this concerns. My suggestion was to log a warning so that developers would know not to use ion-sdk's Ed25519 methods until the nodes support it.

If yes, why weren't this concerns brought up 2+ months ago [when I started to work on this](https://github.com/decentralized-identity/ion-sdk/pull/20)?

thehenrytsai commented:
@gjgd supporting ED25519 would be a great feature add to ION, so thank you for working on this, while I don't understand (still don't) why you are doing all this work for free, but wasn't going to complain! So we'd love to have the code merged in as soon as possible, but in principle not before actual ION node support. Having an SDK that allows people to create operations that are going to be rejected by ION just feels wrong. I don't understand why wouldn't we have support for ED25519 in ION itself before adding support for it in ION-SDK, maybe you can elaborate why you think it should be done in reverse? 

> If yes, why weren't this concerns brought up 2+ months ago https://github.com/decentralized-identity/ion-sdk/pull/20?

My apologies for not making expectations clear in my comment in the previous PR: "The PR looks great overall, given that it really isn't hooked up with anything really, so I see this more as a preparation to more PRs to come".

Basically the PR was approved on the basis that the code was not hooked up to any public facing paths (in fact I thought what you did was clever due to that fact), and by "more PRs" I was imagining both in Sidetree repo as well as ION-SDK repo, and IMO order is important here.

csuwildcat commented:
This thread is confusing to me - I am fully supportive of Ed25519 being added as an operation signing option, but was there a plan to have a PR to make it so it was actually useful to folks? I ask because without a matching PR for the node side of things, it seems like we're handing people something confusing that wouldn't actually work as intended. Can't you provide some clarity on your thinking @gjgd?

OR13 commented:
We'll keep an eye on this repo and wait till you implement support for Ed25519, until then we will use the fork where the feature is available.

We use this library / the fork for 2 other did methods that are not ION, one on Ethereum, and one on Amazon QLDB.

We also don't use the ION codebase for either of those since it bundles bitcoin dependencies.

Its might have been more accurate to call this package a `sidetree-did-method-sdk` since that is how we use it... but I can see the logic in not accepting a PR here when the ION side is not supported.

I can't speak for @gjgd , but this is an open source repo and I think it's fine to close a PR when it's not making progress... or when you feel like the suggested changes are more than you have time to volunteer to do for free.

gjgd commented:
Was planning to use this PR to generate fixtures, that will be used to test a node implementation. I do think this change can stand on its own, but apparently we disagree on that point. 

I hear the concern that the change could be confusing until Ed25519 is implemented on the node side. My suggestion to address this is to:
- Make generating secp256K1 keys a default in [IonKey](https://github.com/decentralized-identity/ion-sdk/blob/0af90f7944fa45a9151d782dd70455d9481ed71f/lib/IonKey.ts#L13) so that developers don't accidentally generate Ed25519 keys for signing Sidetree operations.
- Emit a warning when generating Ed25519 keys until the node implementation is complete.

If this is acceptable, happy to reopen this PR. But to be clear, I'm no longer willing to commit to more work Ed25519 support on the DIF sidetree repo, for the reasons I mentioned on Slack.

csuwildcat commented:
@thehenrytsai @gjgd what if we accepted the PR with the mod to warn, defaulted to secp, and basically buried the alternate key path until we have ION node support without any docs that say it exists? Would this be an acceptable compromise?
